parts/linux/cloud-init/artifacts/ubuntu/cse_install_ubuntu.sh

```diff
@@ -129,42 +129,66 @@ updateAptWithMicrosoftPkg() {
 
 # CSE+VHD can dictate the containerd version, users don't care as long as it works
 installStandaloneContainerd() {
-    CONTAINERD_VERSION=$1
-    CPU_ARCH=$(getCPUArch)  #amd64 or arm64
     # azure-built runtimes have a "+azure" suffix in their version strings (i.e 1.4.1+azure). remove that here.
     CURRENT_VERSION=$(containerd -version | cut -d " " -f 3 | sed 's|v||' | cut -d "+" -f 1)
     CURRENT_COMMIT=$(containerd -version | cut -d " " -f 4)
-    # v1.4.1 is our lowest supported version of containerd
-    
-    # we always default to the .1 patch versons
-    CONTAINERD_PATCH_VERSION="${2:-1}"
+    echo "Before installing containerd, containerd version: ${CURRENT_VERSION}, commit: ${CURRENT_COMMIT}"
 
-    #if there is no containerd_version input from RP, use hardcoded version
-    if [[ -z ${CONTAINERD_VERSION} ]]; then
-        CONTAINERD_VERSION="1.4.9"
-        CONTAINERD_PATCH_VERSION="3"
-        echo "Containerd Version not specified, using default version: ${CONTAINERD_VERSION}-${CONTAINERD_PATCH_VERSION}"
-    else
-        echo "Using specified Containerd Version: ${CONTAINERD_VERSION}-${CONTAINERD_PATCH_VERSION}"
-    fi
-
-    if semverCompare ${CURRENT_VERSION:-"0.0.0"} ${CONTAINERD_VERSION}; then
-        echo "currently installed containerd version ${CURRENT_VERSION} is greater than (or equal to) target base version ${CONTAINERD_VERSION}. skipping installStandaloneContainerd."
-    else
-        echo "installing containerd version ${CONTAINERD_VERSION}"
+    CONTAINERD_PACKAGE_URL="${CONTAINERD_PACKAGE_URL:=}"
+    # the user-defined package url is always picked first, and the others options won't tried when this one fails
+    if [[ ! -z ${CONTAINERD_PACKAGE_URL} ]]; then
+        echo "Installing containerd from user input: ${CONTAINERD_PACKAGE_URL}"
+        # NOTE(mainred): the containerd arch depends on the input from the user
+        mkdir -p $CONTAINERD_DOWNLOADS_DIR
+        CONTAINERD_DEB_TMP=${CONTAINERD_PACKAGE_URL##*/}
+        CONTAINERD_DEB_FILE="$CONTAINERD_DOWNLOADS_DIR/${CONTAINERD_DEB_TMP}"
+        retrycmd_curl_file 120 5 60 ${CONTAINERD_DEB_FILE} ${CONTAINERD_PACKAGE_URL} || exit $ERR_CONTAINERD_DOWNLOAD_TIMEOUT
+        if [[ ! -f "${CONTAINERD_DEB_FILE}" ]]; then
```

Comment:
checking the file right after downloading? just let the download fail?

Reply:
exit with `$ERR_CONTAINERD_DOWNLOAD_TIMEOUT`, though not accurate in case of other failures, won't co-exist with $CONTAINERD_DEB_FILE is not null, right? Let me simplify the code to remove this `if` then.